### PR TITLE
feat: add tenant persistence for multi-tenancy

### PR DESCRIPTION
**Pull request by yu-teo**
https://redhat.atlassian.net/browse/RHOAIENG-62764

## Description
This PR focuses on required changes for the aforementioned S4 step of multi-tenancy implementation.
What has been changed:
- added HeaderTenant = "X-MaaS-Tenant" in `constant/const.go`
-  added Tenant string to UserContext in `token/types.go`
- extract X-MaaS-Tenant header in ExtractUserInfo() (required, 500 if missing) in `token/handler.go`
- added Tenant to ValidationResult in `api_keys/types.go`
- added `tenant` param on CreateAPIKey(), Search(), BulkRevokeAPIKeys(); tenant returned in `ValidateAPIKey` all part of `api_keys/service.go`
- tenant enforcement on all endpoints via isAuthorizedForKey handled in `api_keys/handler.go`
- added tenant param on Search(), InvalidateAll() in api_keys/store_interface.go
- make required changes in `api_keys/store_postgres.go`
- added tenant filtering in filterKeys(), Search() and InvalidateAll() in `api_keys/store_mock.go`
- made a change to the `0004_add_tenant_column.up.sql` by replacing single-column index with composite `(tenant, username, created_at DESC)`

Modified + Created tests:
- `handler_test.go`: tenant on all UserContext/AddKey calls. Also added new cross-tenant isolation tests
- `service_test.go`: tenant param on all CreateAPIKey/BulkRevokeAPIKeys/Search calls + TestValidateAPIKey_ReturnsTenant, TestValidateAPIKey_EmptyTenantReturnsEmpty, TestValidateAPIKey_TenantNotExposedOnInvalid, TestBulkRevokeAPIKeys_TenantScopedCount
- `store_test.go`: tenant param on InvalidateAll calls + TestSearchByTenant, TestInvalidateAll_TenantScoped
- `handlers/models_test.go`: `X-MaaS-Tenant` header on all HTTP test requests + TestExtractUserInfo_TenantHeader

Additionally upon initial reviews:
The gateway-level `buildGatewayAuthPolicySpec` is the single source of truth for auth headers on both API key and OpenShift/OIDC token paths include `X-MaaS-Tenant` for both

## How Has This Been Tested?
Relevant unit-tests have been modified. And new testing for newly-introduced edge cases added as part of this PR.

## Merge criteria:

- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

## Summary by CodeRabbit

* **New Features**
  * Tenant-aware API key management: all key operations are scoped to the caller’s tenant.
* **Security**
  * Mandatory tenant header validation for requests; cross-tenant access is rejected.
  * Tenant included in API key validation results to preserve tenant context.
* **Chores**
  * Database migration added tenant column and indexing to support tenant-scoped lookups.
* **Tests**
  * Expanded unit and end-to-end tests to verify tenant isolation and behavior.

### CODE REVIEW

File: maas-api/internal/api_keys/handler_test.go [hunk 1]
```diff
@@ -1075,7 +1105,9 @@ func TestCreateAPIKey_SubscriptionSelectErrors(t *testing.T) {
 			assert.Equal(t, apiKeySubscriptionResolutionErrCode, resp["code"])
 			assert.Equal(t, apiKeySubscriptionResolutionErrMsg, resp["error"])
 
-			res, err := store.Search(context.Background(), user.Username, &SearchFilters{}, &SortParams{By: DefaultSortBy, Order: DefaultSortOrder}, &PaginationParams{Limit: 10, Offset: 0})
+			res, err := store.Search(context.Background(), user.Username, "",
+				&SearchFilters{}, &SortParams{By: DefaultSortBy, Order: DefaultSortOrder},
+				&PaginationParams{Limit: 10, Offset: 0})
```
[the remark this answers is not in the document]
Comment (yu-teo) [reply maas-api/internal/api_keys/handler_test.go:1110]: The finding is invalid. The interface signature is Search(ctx, username, tenant, ...), not Search(ctx, tenant, username, ...). The call site store.Search(ctx, user.Username, "", ...) has the correct argument order: username first, then tenant (empty string, matching the keys that were never created due to the subscription error).

[the remark this answers is not in the document]
Comment (yu-teo) [reply maas-api/internal/api_keys/handler_test.go:1110]: The finding is invalid. The interface signature is Search(ctx, username, tenant, ...), not Search(ctx, tenant, username, ...). The call site store.Search(ctx, user.Username, "", ...) has the correct argument order: username first, then tenant (empty string, matching the keys that were never created due to the subscription error).

File: maas-api/internal/api_keys/service.go [hunk 2]
```diff
@@ -68,7 +68,7 @@ type CreateAPIKeyResponse struct {
 // Admins can create keys for other users by specifying a different username.
 func (s *Service) CreateAPIKey(
 	ctx context.Context, username string, userGroups []string, name, description string,
-	expiresIn *time.Duration, ephemeral bool, requestedSubscription string,
+	expiresIn *time.Duration, ephemeral bool, requestedSubscription string, tenant string,
 ) (*CreateAPIKeyResponse, error) {
```
[the remark this answers is not in the document]
Comment (yu-teo) [reply maas-api/internal/api_keys/service.go:72]: The finding was about defense-in-depth, but in this case the defense conflicts with backward compatibility. The tenant validation at the service layer would break backward compatibility — existing deployments that haven't configured the `X-MaaS-Tenant` header yet, or legacy code paths that pass empty tenant, would fail hard.
The middleware already enforces non-empty tenant for external API calls (returns 500 if header missing). That's the system boundary where validation belongs. Adding a second check in the service layer means even internal/programmatic callers (future batch jobs, migration scripts, etc.) can't create keys with empty tenant, which blocks the transition path.

File: maas-api/internal/api_keys/service.go [hunk 3]
```diff
@@ -143,7 +143,7 @@ func (s *Service) CreateAPIKey(
 	// Note: prefix is NOT stored (security - reduces brute-force attack surface)
 	// userGroups stored as PostgreSQL TEXT[] array (no JSON marshaling needed)
 	// Hash is SHA-256(key_id + secret) where key_id is embedded in the API key as per-key salt
-	if err := s.store.AddKey(ctx, username, keyID, hash, name, description, userGroups, subscriptionName, "", &expiresAt, ephemeral); err != nil {
+	if err := s.store.AddKey(ctx, username, keyID, hash, name, description, userGroups, subscriptionName, tenant, &expiresAt, ephemeral); err != nil {
```
[the remark this answers is not in the document]
Comment (yu-teo) [reply maas-api/internal/api_keys/service.go:146]: Empty tenant must remain valid for backward compatibility: existing keys in the database have `tenant = ''` from migration `0004's DEFAULT ''`, and the TestValidateAPIKey_EmptyTenantReturnsEmpty test explicitly verifies these legacy keys still validate. Tenant enforcement happens at the middleware boundary (ExtractUserInfo rejects missing/empty header), not at the store layer. The DB column already has `NOT NULL DEFAULT ''` — adding a non-empty constraint would break the migration for existing rows.

File: maas-api/internal/api_keys/service.go [hunk 4]
```diff
@@ -279,20 +280,21 @@ func (s *Service) RevokeAPIKey(ctx context.Context, keyID string) error {
...
 	ctx context.Context,
 	username string,
+	tenant string,
 	filters *SearchFilters,
 	sort *SortParams,
 	pagination *PaginationParams,
 ) (*PaginatedResult, error) {
-	return s.store.Search(ctx, username, filters, sort, pagination)
+	return s.store.Search(ctx, username, tenant, filters, sort, pagination)
 }
```
[the remark this answers is not in the document]
Comment (yu-teo) [reply maas-api/internal/api_keys/service.go:289]: Tenant is validated at the middleware boundary (ExtractUserInfo rejects empty/missing header before any handler runs). The service layer intentionally accepts empty tenant for backward compatibility with legacy keys. All three findings (CreateAPIKey, AddKey, Search) propose the same guard that was already evaluated and rejected.

File: maas-api/internal/api_keys/service.go [hunk 5]
```diff
@@ -279,20 +280,21 @@ func (s *Service) RevokeAPIKey(ctx context.Context, keyID string) error {
...
 // BulkRevokeAPIKeys revokes all active keys for a user
 // Returns count of revoked keys.
-func (s *Service) BulkRevokeAPIKeys(ctx context.Context, username string) (int, error) {
+func (s *Service) BulkRevokeAPIKeys(ctx context.Context, username string, tenant string) (int, error) {
 	if username == "" {
 		return 0, errors.New("username is required")
 	}
-	return s.store.InvalidateAll(ctx, username)
+	return s.store.InvalidateAll(ctx, username, tenant)
 }
```
[the remark this answers is not in the document]
Comment (yu-teo) [reply maas-api/internal/api_keys/service.go:298]: Tenant is enforced at the middleware boundary, and empty tenant must remain valid at the service layer for backward compatibility with legacy keys.

File: maas-api/internal/token/handler.go [hunk 6]
```diff
@@ -106,10 +106,25 @@ func (h *Handler) ExtractUserInfo() gin.HandlerFunc {
...
+				"header", constant.HeaderTenant,
+			)
+			c.JSON(http.StatusInternalServerError, gin.H{
+				"error":         "Exception thrown while generating token",
+				"exceptionCode": "AUTH_FAILURE",
+				"refId":         "004",
+			})
+			c.Abort()
+			return
+		}
```
[the remark this answers is not in the document]
Comment (yu-teo) [reply maas-api/internal/token/handler.go:121]: The tenant header is set by the auth policy (Authorino), not by end users. The existing headers (X-MaaS-Username, X-MaaS-Group) don't validate length or character sets either. Adding regex/length validation to tenant alone would be inconsistent, and the auth policy is the trusted source — it controls what values reach this header. This is an internal system boundary, not a user-facing input boundary.

**Comment by jrhyness** [comment]
How does the header get set?   Will it work for both apikey and openshift tokens?

**Comment by yu-teo** [comment]
> How does the header get set? Will it work for both apikey and openshift tokens?

For API key path `ValidationResult()` already returns `tenant`, so we add a header entry sourced from `auth.metadata.apiKeyValidation.tenant`. I did not account for OpenShift tokens. Checking now.

**Comment by yu-teo** [comment]
@coderabbitai review

**Comment by yu-teo** [comment]
@coderabbitai review

**Comment by yu-teo** [comment]
> Caution
> 
> Some comments are outside the diff and can’t be posted inline due to platform limitations.
> 
> ⚠️ Outside diff range comments (1)
> > maas-controller/pkg/platform/tenantreconcile/params.go (1)> `62-113`: _⚠️ Potential issue_ | _🔴 Critical_ | _⚡ Quick win_
> > > **Wire `patchMaaSAPIAuthPolicy` into `applyPlatformParams`.**
> > > This patch is dead code today. `applyPlatformParams` never dispatches AuthPolicy resources, so `deployment/base/maas-api/policies/auth-policy.yaml` Line 44 keeps the `.placehold.` validation URL and Line 114 keeps `TENANT_PLACEHOLDER`. That breaks API-key auth and mis-tags OpenShift requests with the wrong tenant, which is an authorization boundary failure (CWE-285). Add an AuthPolicy case in the switch and invoke `patchMaaSAPIAuthPolicy` there.
> > > Also applies to: 182-228
> > > 
> > > 🤖 Prompt for AI Agents
> > > ```
> > > Verify each finding against current code. Fix only still-valid issues, skip the
> > > rest with a brief reason, keep changes minimal, and validate.
> > > 
> > > In `@maas-controller/pkg/platform/tenantreconcile/params.go` around lines 62 -
> > > 113, applyPlatformParams never handles AuthPolicy resources so
> > > patchMaaSAPIAuthPolicy is dead; add a switch case that matches the AuthPolicy
> > > GVK (GVKAuthPolicy) and the MaaS API auth policy resource name (e.g.,
> > > MaaSAPIAuthPolicyName) and invoke patchMaaSAPIAuthPolicy(log, r, params)
> > > returning any error. Update both places of the switch (the current block and the
> > > later block referenced by the reviewer) so AuthPolicy resources are dispatched
> > > the same way as other resources.
> > > ```
> 
> 🤖 Prompt for all review comments with AI agents
> ```
> Verify each finding against current code. Fix only still-valid issues, skip the
> rest with a brief reason, keep changes minimal, and validate.
> 
> Outside diff comments:
> In `@maas-controller/pkg/platform/tenantreconcile/params.go`:
> - Around line 62-113: applyPlatformParams never handles AuthPolicy resources so
> patchMaaSAPIAuthPolicy is dead; add a switch case that matches the AuthPolicy
> GVK (GVKAuthPolicy) and the MaaS API auth policy resource name (e.g.,
> MaaSAPIAuthPolicyName) and invoke patchMaaSAPIAuthPolicy(log, r, params)
> returning any error. Update both places of the switch (the current block and the
> later block referenced by the reviewer) so AuthPolicy resources are dispatched
> the same way as other resources.
> ```
> 
> ℹ️ Review info
> ⚙️ Run configuration
> **Configuration used**: Central YAML (base), Organization UI (inherited)
> 
> **Review profile**: CHILL
> 
> **Plan**: Enterprise
> 
> **Run ID**: `4df23acc-0cc4-4d28-b1ce-e02f7d34f820`
> 
> 📥 Commits
> Reviewing files that changed from the base of the PR and between [4111553](https://github.com/opendatahub-io/models-as-a-service/commit/411155317b69794c0b643dc58bdb2298c1b6f7f4) and [d4fe0f4](https://github.com/opendatahub-io/models-as-a-service/commit/d4fe0f48a7975146e2f07b3813f7904cefb0553f).
> 
> 📒 Files selected for processing (17)
> * `deployment/base/maas-api/policies/auth-policy.yaml`
> * `maas-api/db/schema/0004_add_tenant_column.up.sql`
> * `maas-api/internal/api_keys/handler.go`
> * `maas-api/internal/api_keys/handler_test.go`
> * `maas-api/internal/api_keys/service.go`
> * `maas-api/internal/api_keys/service_test.go`
> * `maas-api/internal/api_keys/store_interface.go`
> * `maas-api/internal/api_keys/store_mock.go`
> * `maas-api/internal/api_keys/store_postgres.go`
> * `maas-api/internal/api_keys/store_test.go`
> * `maas-api/internal/api_keys/types.go`
> * `maas-api/internal/constant/const.go`
> * `maas-api/internal/handlers/models_test.go`
> * `maas-api/internal/token/handler.go`
> * `maas-api/internal/token/handler_test.go`
> * `maas-api/internal/token/types.go`
> * `maas-controller/pkg/platform/tenantreconcile/params.go`
> 
> 🚧 Files skipped from review as they are similar to previous changes (14)

@coderabbitai The auth-policy.yaml that patchMaaSAPIAuthPolicy would patch has been intentionally removed from Kustomize rendering (resources: [] in deployment/base/maas-api/policies/kustomization.yaml). Auth is now handled by the gateway-level policy built by maasauthpolicy_controller.go, which already has tenant headers. Wiring patchMaaSAPIAuthPolicy into applyPlatformParams would add code to patch a resource that is never rendered — it would never execute and just add dead surface area

File: maas-controller/pkg/platform/tenantreconcile/params.go [hunk 7]
```diff
@@ -179,6 +179,55 @@ func patchHTTPRoute(log logr.Logger, r *unstructured.Unstructured, params Platfo
 	return nil
 }
 
+func patchMaaSAPIAuthPolicy(log logr.Logger, r *unstructured.Unstructured, params PlatformParams) error {
```
Comment (jrhyness) [inline maas-controller/pkg/platform/tenantreconcile/params.go]: This function is obsolete.  It was removed in PR 912.

File: deployment/base/maas-api/policies/auth-policy.yaml [hunk 8]
```diff
@@ -97,6 +97,23 @@ spec:
               selector: auth.identity.user.groups.@tostr
             key: X-MaaS-Group
             priority: 1
+          # Tenant: from API key validation (when API key used)
```
Comment (jrhyness) [inline deployment/base/maas-api/policies/auth-policy.yaml]: I don't think this file is used any more.   You could revert this change.

**Review by jrhyness — CHANGES_REQUESTED** [review]
Looks good.   I left a couple comments.  This is due to the gateway-level auth policies.

File: maas-controller/pkg/platform/tenantreconcile/params.go (hunk 7 above)
Comment (yu-teo) [reply maas-controller/pkg/platform/tenantreconcile/params.go]: Removed, I must have brought it in a merge-conflict... Thank you for spotting it. Cleaning up now.

File: deployment/base/maas-api/policies/auth-policy.yaml (hunk 8 above)
Comment (yu-teo) [reply deployment/base/maas-api/policies/auth-policy.yaml]: Ok, so if we do not use this file... Am I correct to say that the tenant header for the gateway-level policy is correctly handled in `maasauthpolicy_controller.go` via `buildGatewayAuthPolicySpec`?

File: maas-controller/pkg/controller/maas/maasauthpolicy_controller.go [hunk 9]
```diff
@@ -777,6 +777,33 @@ allow {
...
+					},
+					"X-MaaS-Tenant-Token": map[string]any{
+						"when": []any{
+							map[string]any{
+								"predicate": `!request.headers.authorization.startsWith("Bearer sk-oai-")`,
+							},
+						},
+						"plain": map[string]any{
+							"expression": fmt.Sprintf(`"%s"`, r.MaaSAPINamespace),
+						},
```
[the remark this answers is not in the document]
Comment (yu-teo) [reply maas-controller/pkg/controller/maas/maasauthpolicy_controller.go:803]: Great finding, thank you. I will fix it with another commit.

File: deployment/base/maas-api/policies/auth-policy.yaml (hunk 8 above)
Comment (jrhyness) [reply deployment/base/maas-api/policies/auth-policy.yaml]: Yes, that all looks good and should handle everything.

**Review by jrhyness — DISMISSED** [review]
lgtm

File: test/e2e/tests/test_models_endpoint.py [hunk 10]
```diff
@@ -73,20 +73,26 @@
 
 
 def _request_with_gateway_retry(method, url, retries=GATEWAY_PROPAGATION_RETRIES, **kwargs):
```
Comment (yu-teo) [inline test/e2e/tests/test_models_endpoint.py:75]: I have encountered an error `refId: "001"` which means `ExtractUserInfo()` middleware didn't find the `X-MaaS-Username` header. My guess is that it isn't a tenant issue but rather an `Authorino` propagation timing issue.
How it happens:
1. The test creates an API key, waits for reconciliation, then immediately calls `/v1/models`
2. The request goes through the gateway -> Authorino -> maas-api
3. Authorino needs to evaluate the `AuthPolicy`, call the `/internal/v1/api-keys/validate` callback, and inject the response headers (`X-MaaS-Username`, `X-MaaS-Group`, `X-MaaS-Tenant`)
4. If Authorino hasn't finished loading/caching the `AuthPolicy` yet, it may pass the request through without injecting headers — causing the `500` error.

So the proposed fix would be:
Replace `requests.get()` with `_request_with_gateway_retry()` (which mathces the pattern used for other tests in the file). The `_request_with_gateway_retry` helper handles the timeout and verify params internally, so those get dropped from the call.

File: test/e2e/tests/test_models_endpoint.py [hunk 11]
```diff
@@ -73,20 +73,26 @@
...
+        is_empty_403 = r.status_code == 403 and not r.text.strip()
+        is_auth_propagation_500 = (r.status_code == 500
+                                   and "AUTH_FAILURE" in r.text)
+        if (is_empty_403 or is_auth_propagation_500) and attempt < retries:
+            log.info(f"Gateway not ready (HTTP {r.status_code}, attempt {attempt}/{retries}), "
+                     f"retrying in {GATEWAY_PROPAGATION_DELAY}s...")
+            time.sleep(GATEWAY_PROPAGATION_DELAY)
+            continue
         return r
     return r  # last attempt's response — assertion will catch the failure
```
[the remark this answers is not in the document]
Comment (yu-teo) [reply test/e2e/tests/test_models_endpoint.py:97]: The retry only fires on transient gateway propagation errors (empty 403, AUTH_FAILURE 500) — these mean the request never reached the application layer, so no side effects occurred. There's nothing to duplicate. Adding idempotency guards would add complexity for a scenario that can't cause harm.

**Review by jrhyness — APPROVED** [review]
lgtm
